pkg/services/ngalert/store/instance_database.go

```diff
@@ -150,14 +191,86 @@ func (st DBstore) FetchOrgIds(ctx context.Context) ([]int64, error) {
 	return orgIds, err
 }
 
-func (st DBstore) DeleteAlertInstance(ctx context.Context, orgID int64, ruleUID, labelsHash string) error {
-	return st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		_, err := sess.Exec("DELETE FROM alert_instance WHERE rule_org_id = ? AND rule_uid = ? AND labels_hash = ?", orgID, ruleUID, labelsHash)
+// DeleteAlertInstances deletes instances with the provided keys in a single transaction.
+func (st DBstore) DeleteAlertInstances(ctx context.Context, keys ...models.AlertInstanceKey) error {
+	if len(keys) == 0 {
+		return nil
+	}
+
+	type data struct {
+		ruleOrgID   int64
+		ruleUID     string
+		labelHashes []interface{}
+	}
+
+	maxRows := 200
+	rowData := data{
+		0, "", make([]interface{}, 0, maxRows),
+	}
+	placeholderArray := strings.Builder{}
+	placeholderArray.WriteString("(")
+
+	execQuery := func(rd data, s *sqlstore.DBSession) error {
+		if len(rd.labelHashes) == 0 {
+			return nil
+		}
+
+		hashPlaceHolders := placeholderArray.String()
+		hashPlaceHolders = strings.TrimRight(hashPlaceHolders, ", ")
+		hashPlaceHolders = hashPlaceHolders + ")"
+
+		queryString := fmt.Sprintf(
+			"DELETE FROM alert_instance WHERE rule_org_id = ? AND rule_uid = ? AND labels_hash IN %s;",
+			hashPlaceHolders,
+		)
+
+		execArgs := make([]interface{}, 0, 3+len(rd.labelHashes))
+		execArgs = append(execArgs, queryString, rd.ruleOrgID, rd.ruleUID)
+		execArgs = append(execArgs, rd.labelHashes...)
+		_, err := s.Exec(execArgs...)
 		if err != nil {
 			return err
 		}
+
+		return nil
+	}
+
+	err := st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		counter := 0
+		for _, k := range keys {
+			counter++
+			// When a rule ID changes or we hit 200 hashes, issue a statement.
+			if rowData.ruleOrgID != k.RuleOrgID || rowData.ruleUID != k.RuleUID || len(rowData.labelHashes) >= 200 {
+				err := execQuery(rowData, sess)
```

Comment:
Worst case, where `keys` are in random order, this will execute one query per key? Should we not group `keys` by `RuleOrgID` and `RuleUID` before executing the query?